### PR TITLE
Resolve UB in Arc/Weak interaction (2)

### PR DESCRIPTION
**Pull request by Diggsey**
Use raw pointers to avoid making any assertions about the data field.

Follow up from #72479, see that PR for more detail on the motivation.

@RalfJung I was able to avoid a lot of the changes to `Weak`, by making a helper type (`WeakInner`) - because of auto-deref and because the fields have the same name, the rest of the code continues to compile.

### CODE REVIEW

File: src/liballoc/sync.rs [hunk 1]
```diff
@@ -866,12 +866,10 @@ impl<T: ?Sized> Arc<T> {
...
         // allocation itself (there may still be weak pointers lying around).
-        ptr::drop_in_place(&mut self.ptr.as_mut().data);
+        ptr::drop_in_place(Self::get_mut_unchecked(self));
 
-        if self.inner().weak.fetch_sub(1, Release) == 1 {
-            acquire!(self.inner().weak);
-            Global.dealloc(self.ptr.cast(), Layout::for_value(self.ptr.as_ref()))
-        }
+        // Convert ourselves into a weak pointer and drop it
+        Weak { ptr: self.ptr };
```
Comment (RalfJung) [inline src/liballoc/sync.rs:872]: Uh, is this change needed? Feels mostly orthogonal. I'd prefer to keep such tricky PRs as small and focused as possible.

File: src/liballoc/sync.rs [hunk 2]
```diff
@@ -1277,7 +1275,7 @@ impl<T: ?Sized> Arc<T> {
     #[inline]
     #[unstable(feature = "get_mut_unchecked", issue = "63292")]
     pub unsafe fn get_mut_unchecked(this: &mut Self) -> &mut T {
-        &mut this.ptr.as_mut().data
+        &mut (*this.ptr.as_ptr()).data
```
Comment (RalfJung) [inline src/liballoc/sync.rs:1278]: Maybe add a comment saying that we carefully do *not* create a reference covering the "count" fields.

File: src/liballoc/sync.rs [hunk 3]
```diff
@@ -1673,8 +1678,15 @@ impl<T: ?Sized> Weak<T> {
...
     #[inline]
-    fn inner(&self) -> Option<&ArcInner<T>> {
-        if is_dangling(self.ptr) { None } else { Some(unsafe { self.ptr.as_ref() }) }
+    fn inner(&self) -> Option<WeakInner<'_>> {
+        if is_dangling(self.ptr) {
+            None
+        } else {
+            Some(unsafe {
+                let ptr = self.ptr.as_ptr();
+                WeakInner { strong: &(*ptr).strong, weak: &(*ptr).weak }
```
Comment (RalfJung) [inline src/liballoc/sync.rs:1687]: Same here about reference creation.

File: src/liballoc/sync.rs [hunk 4]
```diff
@@ -1568,6 +1566,13 @@ impl<T> Weak<T> {
     }
 }
 
+// Helper type to allow accessing the reference counts without
+// making any assertions about the data field.
```
Comment (RalfJung) [inline src/liballoc/sync.rs:1570]:
```suggestion
/// Helper type to allow accessing the reference counts without
/// making any assertions about the data field.
```

**Comment by RalfJung** [comment]
> I was able to avoid a lot of the changes to Weak, by making a helper type (WeakInner) - because of auto-deref and because the fields have the same name, the rest of the code continues to compile.

That's a very clever trick, I like it. :)

File: src/liballoc/sync.rs (hunk 1 above)
Comment (Diggsey) [reply src/liballoc/sync.rs:872]: I suppose not, but it does de-duplicate some fairly tricky logic, and with the change to Weak::drop that this PR makes, these two implementations would otherwise deviate. If you feel strongly I can revert it, but it seems a little silly.

File: src/liballoc/sync.rs (hunk 2 above)
Comment (Diggsey) [reply src/liballoc/sync.rs:1278]: 👍

File: src/liballoc/sync.rs (hunk 4 above)
Comment (Diggsey) [reply src/liballoc/sync.rs:1570]: Does this make a difference for private types?

File: src/liballoc/sync.rs (hunk 3 above)
Comment (Diggsey) [reply src/liballoc/sync.rs:1687]: 👍

File: src/liballoc/sync.rs (hunk 4 above)
Comment (RalfJung) [reply src/liballoc/sync.rs]: When working in an IDE I assume doc comments are picked up for private types.

You'll notice that many private functions also have doc comments.

File: src/liballoc/sync.rs [hunk 5]
```diff
@@ -1277,7 +1275,9 @@ impl<T: ?Sized> Arc<T> {
     #[inline]
     #[unstable(feature = "get_mut_unchecked", issue = "63292")]
     pub unsafe fn get_mut_unchecked(this: &mut Self) -> &mut T {
-        &mut this.ptr.as_mut().data
+        // We are careful to *not* create a reference covering the "count" fields, as
+        // this would alias with concurrent access to the reference counts.
```
Comment (RalfJung) [inline src/liballoc/sync.rs:1279]:
```suggestion
        // this would alias with concurrent access to the reference counts (e.g. by `Weak`).
```

File: src/liballoc/sync.rs (hunk 1 above)
Comment (RalfJung) [reply src/liballoc/sync.rs:872]: Hm, makes sense. You now have an extra `if` here but this is already called `drop_slow` so it's probably not a big deal.

Two comments:
* "convert ourselves into weak" sounds like a `fn(self) -> Weak` kind of situation, but that's not what this is. So I'd rather say "drop the weak count held by all strong references" or so.
* When dropping is the core part of what happens, I think it is better to write `drop(Weak { ... });`, to make that clear and to also make it so that one can see clearly where the drop happens.

Comment (Diggsey) [reply src/liballoc/sync.rs:872]: - Updated the comment.
- Yep, also I would expect this to check to be trivially optimised out - our previous access to `ptr` implicitly asserts that it's a valid pointer.

File: src/liballoc/sync.rs [hunk 6]
```diff
@@ -866,12 +866,10 @@ impl<T: ?Sized> Arc<T> {
...
         // allocation itself (there may still be weak pointers lying around).
-        ptr::drop_in_place(&mut self.ptr.as_mut().data);
+        ptr::drop_in_place(Self::get_mut_unchecked(self));
 
-        if self.inner().weak.fetch_sub(1, Release) == 1 {
-            acquire!(self.inner().weak);
-            Global.dealloc(self.ptr.cast(), Layout::for_value(self.ptr.as_ref()))
-        }
+        // Drop the weak ref collectively held by all strong references
+        Weak { ptr: self.ptr };
```
Comment (RalfJung) [inline src/liballoc/sync.rs:872]:
```suggestion
        // Drop the weak ref collectively held by all strong references.
        drop(Weak { ptr: self.ptr });
```

Comment (RalfJung) [reply src/liballoc/sync.rs]: (My suggestion also added the missing `.` to the end of the sentence but whatever.^^)

**Review by RalfJung — APPROVED** [review]
LGTM!

I'd prefer if someone else would also review this --the more eyes, the better.

**Review by dtolnay — APPROVED** [review]
Looks good to me as well. Thanks for the helpful round of review.
